armv7 CI failures with recent serialization changes

### Description

There are currently CI failures on armv7 following (I think) https://github.com/astropy/astropy/pull/18703 - @taldcroft  would you be able to take a look? (if you open a PR to fix this, add the Extra CI label - thanks!)

```
______________ [doctest] astropy.table.connect.represent_indices _______________
  288 
  289 The format is most easily illustrated via the ECSV output in this example::
  290 
  291   >>> from astropy.table import QTable
  292   >>> import sys
  293   >>> t = QTable()
  294   >>> t["a"] = [2, 3, 1]
  295   >>> t["b"] = [3, 5, 4]
  296   >>> t.add_index("a")
  297   >>> t.write(sys.stdout, format="ecsv", write_indices=True)
  Differences (unified diff with -expected +actual):
      @@ -3,11 +3,11 @@
       # datatype:
       # - name: a
      -#   datatype: int64
      +#   datatype: int32
       #   meta: !!omap
       #   - __indices__:
       #     - colnames: [a]
       #       index_colname: __index__
      -# - {name: b, datatype: int64}
      -# - {name: __index__, datatype: int64}
      +# - {name: b, datatype: int32}
      +# - {name: __index__, datatype: int32}
          '#       index_colname: __index__',
    -     '# - {name: __index__, datatype: int64}',
    ?                                         ^^
    +     '# - {name: __index__, datatype: int32}',
    ?                                         ^^
          '# schema: astropy-2.0',
          'a __index__',
          '1 0',
          '3 2',
          '2 1',
      ]
  FAILED astropy/table/tests/test_index.py::test_indices_serialization_representation_multiple - AssertionError: assert ['# %ECSV 1.0... !!omap', ...] == ['# %ECSV 1.0... !!omap', ...]
    
    At index 4 diff: '#   datatype: int32' != '#   datatype: int64'
    
    Full diff:
      [
          '# %ECSV 1.0',
          '# ---',
          '# datatype:',
          '# - name: a',
    -     '#   datatype: int64',
    ?                       ^^
    +     '#   datatype: int32',
    ?                       ^^
          '#   meta: !!omap',
          '#   - __indices__:',
          '#     - colnames: [a, __index__1]',
          '#       index_colname: __index__',
          '#       primary: true',
          '#     - colnames: [a]',
          '#       index_colname: __index__2',
    -     '# - {name: __index__1, datatype: int64}',
    ?                                          ^^
    +     '# - {name: __index__1, datatype: int32}',
    ?                                          ^^
    -     '# - {name: __index__, datatype: int64}',
    ?                                         ^^
    +     '# - {name: __index__, datatype: int32}',
    ?                                         ^^
    -     '# - {name: __index__2, datatype: int64}',
    ?                                          ^^
    +     '# - {name: __index__2, datatype: int32}',
    ?                                          ^^
          '# schema: astropy-2.0',
          'a __index__1 __index__ __index__2',
          '1 5 0 0',
          '3 4 2 2',
          '2 3 1 1',
      ]
  = 5 failed, 26782 passed, 3899 skipped, 72 deselected, 214 xfailed in 2455.83s (0:40:55) =
  Error: The process '/home/runner/work/_actions/uraimo/run-on-arch-action/d94c13912ea685de38fccc1109385b83fd79427d/src/run-on-arch.sh' failed with exit code 1
```

### Expected behavior

_No response_

### How to Reproduce

1. Get package from '...'
2. Then run '...'
3. An error occurs.

```python
# Put your Python code snippet here.
```


### Versions

```python
import astropy
astropy.system_info()
```
```
# Paste the result here
```

--- a/astropy/table/connect.py
+++ b/astropy/table/connect.py
@@ -277,32 +277,7 @@ def represent_indices(tbl: Table, /) -> Table:
     automatically serializes column meta into YAML comments, while table meta is assumed
     to be FITS keyword files.
 
-    The format is most easily illustrated via the ECSV output in this example::
-
-      >>> from astropy.table import QTable
-      >>> import sys
-      >>> t = QTable()
-      >>> t["a"] = [2, 3, 1]
-      >>> t["b"] = [3, 5, 4]
-      >>> t.add_index("a")
-      >>> t.write(sys.stdout, format="ecsv", write_indices=True)
-      # %ECSV 1.0
-      # ---
-      # datatype:
-      # - {name: a, datatype: int64}
-      # - {name: b, datatype: int64}
-      # - {name: __index__, datatype: int64}
-      # meta: !!omap
-      # - __table_indices__:
-      #     indices:
-      #     - colnames: [a]
-      #       index_colname: __index__
-      #     primary_key: [a]
-      # schema: astropy-2.0
-      a b __index__
-      2 3 2
-      3 5 0
-      1 4 1
+    See the ``test_indices_serialization_*` tests in ``test_index.py`` for examples.
 
     Parameters
     ----------

--- a/astropy/table/tests/test_index.py
+++ b/astropy/table/tests/test_index.py
@@ -21,6 +21,7 @@ from astropy.utils.exceptions import AstropyDeprecationWarning, AstropyWarning
 from .test_table import SetupData
 
 available_engines = [BST, SortedArray]
+NATIVE_INT_NAME = np.array(0).dtype.name
 
 if HAS_SORTEDCONTAINERS:
     available_engines.append(SCEngine)
@@ -761,8 +762,8 @@ def test_indices_read_unknown_engine():
         "# %ECSV 1.0",
         "# ---",
         "# datatype:",
-        "# - {name: a, datatype: int64}",
-        "# - {name: __index__, datatype: int64}",
+        f"# - {{name: a, datatype: {NATIVE_INT_NAME}}}",
+        f"# - {{name: __index__, datatype: {NATIVE_INT_NAME}}}",
         "# meta: !!omap",
         "# - __table_indices__:",
         "#     indices:",
@@ -800,8 +801,8 @@ def test_indices_serialization_unique_representation():
         "# %ECSV 1.0",
         "# ---",
         "# datatype:",
-        "# - {name: a, datatype: int64}",
-        "# - {name: __index__, datatype: int64}",
+        f"# - {{name: a, datatype: {NATIVE_INT_NAME}}}",
+        f"# - {{name: __index__, datatype: {NATIVE_INT_NAME}}}",
         "# meta: !!omap",
         "# - __table_indices__:",
         "#     indices:",
@@ -834,8 +835,8 @@ def test_indices_serialization_representation_single(engine):
         "# %ECSV 1.0",
         "# ---",
         "# datatype:",
-        "# - {name: a, datatype: int64}",
-        "# - {name: __index__, datatype: int64}",
+        f"# - {{name: a, datatype: {NATIVE_INT_NAME}}}",
+        f"# - {{name: __index__, datatype: {NATIVE_INT_NAME}}}",
         "# meta: !!omap",
         "# - __table_indices__:",
         "#     indices:",
@@ -872,10 +873,10 @@ def test_indices_serialization_representation_multiple():
         "# %ECSV 1.0",
         "# ---",
         "# datatype:",
-        "# - {name: a, datatype: int64}",
-        "# - {name: __index__1, datatype: int64}",
-        "# - {name: __index__, datatype: int64}",
-        "# - {name: __index__2, datatype: int64}",
+        f"# - {{name: a, datatype: {NATIVE_INT_NAME}}}",
+        f"# - {{name: __index__1, datatype: {NATIVE_INT_NAME}}}",
+        f"# - {{name: __index__, datatype: {NATIVE_INT_NAME}}}",
+        f"# - {{name: __index__2, datatype: {NATIVE_INT_NAME}}}",
         "# meta: !!omap",
         "# - __table_indices__:",
         "#     indices:",
@@ -891,6 +892,22 @@ def test_indices_serialization_representation_multiple():
         "2 3 1 1",
     ]
     assert out.getvalue().splitlines() == exp
+
+
+@pytest.mark.parametrize("dtype", [np.int16, np.float32, np.int64, np.float64])
+def test_indices_roundtrip_various_dtypes(dtype):
+    """Test that serialization round-trip works for various index dtypes."""
+    t = Table()
+    t["a"] = np.array([1, 3, 2], dtype=dtype)
+    t["b"] = np.array([5, 6, 7], dtype=dtype)
+    t.add_index("a")
+    t.add_index(["a", "b"])
+    out = io.StringIO()
+    t.write(out, format="ecsv", write_indices=True)
+    t2 = Table.read(out.getvalue(), format="ecsv")
+
+    assert_tables_equal(t, t2)
+    assert_indices_equal(t, t2, [("a",), ("a", "b")])
 
 
 @pytest.mark.parametrize("single_index", [True, False])
@@ -931,6 +948,11 @@ def test_indices_roundtrip_through_file(single_index, fmt, engine, tmp_path):
         # FITS does not round-trip the format
         t2["a"].format = "cxcsec"
 
+    assert_tables_equal(t, t2)
+    assert_indices_equal(t, t2, indices_colnames)
+
+
+def assert_indices_equal(t, t2, indices_colnames):
     assert len(t.indices) == len(t2.indices)
     assert t.primary_key == t2.primary_key
 

